### ISSUE TITLE
Add Cox proportional hazards loss to support DeepSurv models

### ISSUE THREAD

**Issue by salernos** [jtleek/scorcher#26]
### Motivation

It would be useful for `scorcher` to support deep survival models, beginning with **DeepSurv / neural Cox proportional hazards models**.

Most of the functionality needed for DeepSurv is already available through `scorcher` and `torch`. A DeepSurv model is simply a neural network that outputs a single, unrestricted log-risk score and is trained using the Cox partial likelihood. Therefore, we likely do not need a dedicated `DeepSurv` model class or substantial changes to the existing architecture API.

Instead, the main missing component is a **Cox proportional hazards loss function** that can be supplied to `compile_scorch()`.

### Proposed functionality

Add a function such as:

```r
scorch_loss_coxph(
  ties = "efron",
  reduction = "mean"
)
```

The loss would expect:

* `prediction`: an `n x 1` tensor containing predicted log-risk scores
* `target`: an `n x 2` tensor containing:

  1. observed follow-up time
  2. event indicator

For example:

```r
y <- cbind(
  time = time,
  event = event
)

dl <- scorch_create_dataloader(
  input = x,
  output = y,
  batch_size = nrow(x)
)
```

A DeepSurv model could then be defined using the existing scorcher API:

```r
model <- initiate_scorch(dl) |>
  scorch_input("x") |>
  scorch_layer(
    linear,
    in_features = p,
    out_features = 64
  ) |>
  scorch_layer(relu) |>
  scorch_dropout(p = 0.2) |>
  scorch_layer(
    linear,
    in_features = 64,
    out_features = 32
  ) |>
  scorch_layer(relu) |>
  scorch_layer(
    linear,
    in_features = 32,
    out_features = 1
  ) |>
  scorch_output("risk") |>
  compile_scorch(
    loss_fn = scorch_loss_coxph()
  )
```

The final output layer should remain linear, with no sigmoid or other activation.

### Cox partial likelihood

For predicted log-risk scores \(\eta_i = f_\theta(x_i)\), the loss should minimize the negative Cox partial log-likelihood:

$$
-\sum_{i:\delta_i=1}
\left[
\eta_i -
\log\left(
\sum_{j:T_j \geq T_i} e^{\eta_j}
\right)
\right].
$$

This can be implemented entirely using differentiable `torch` operations. After sorting individuals by decreasing observed time, the risk-set denominator can be computed efficiently using `torch_logcumsumexp()`.

A simplified no-ties/Breslow implementation would look conceptually like:

```r
cox_loss <- function(prediction, target) {

  time <- target[, 1]
  event <- target[, 2]

  eta <- prediction$squeeze(2)

  idx <- torch_argsort(
    time,
    descending = TRUE
  )

  eta <- eta[idx]
  event <- event[idx]

  log_risk <- torch_logcumsumexp(
    eta,
    dim = 1
  )

  partial_ll <- event * (eta - log_risk)

  -partial_ll$sum() / event$sum()
}
```

A better implementation should handle tied event times explicitly.

### Tied event times

Initial support could include:

```r
ties = "breslow"
ties = "efron"
```

with `"efron"` as the default to align with `survival::coxph()`.

An exact partial likelihood does not seem necessary for the first implementation.

### Batching

One important consideration is that the Cox partial likelihood depends on risk sets across observations.

Using the current default minibatch training would calculate risk sets only within each batch and therefore would not be equivalent to the standard full-data Cox partial likelihood.

For the initial implementation, documentation should recommend:

```r
batch_size = nrow(x)
```

for standard Cox/DeepSurv training.

It may also be useful for `scorch_loss_coxph()` or the training machinery to warn users when the Cox loss is used with minibatches.

Alternative minibatch or sampled-risk-set approaches could be considered separately in the future.

### Validation

The implementation should be validated against `survival::coxph()`.

A useful test case would use a scorcher model with no hidden layers:

```text
X -> linear(1) -> Cox loss
```

and compare the resulting coefficient estimates and partial log-likelihood with:

```r
survival::coxph(
  survival::Surv(time, event) ~ x1 + x2 + ...
)
```

Tests should cover at least:

* no tied event times
* Breslow ties
* Efron ties
* censoring
* numerical stability for large predicted risk scores
* gradient propagation through the loss
* CPU and, where feasible, GPU execution

### Possible follow-up functionality

These do not need to be part of the initial implementation, but could follow naturally:

```r
scorch_surv(time, event)
```

as a convenience wrapper for creating survival targets.

Additional prediction methods could also eventually support:

```r
predict_scorch(model, new_data, type = "risk")
predict_scorch(model, new_data, type = "survival", times = ...)
```

The latter would require estimating the baseline cumulative hazard after fitting.

### Suggested scope for this issue

For the initial PR, the minimal useful scope would be:

* [ ] Implement `scorch_loss_coxph()`
* [ ] Support Breslow and Efron handling of ties
* [ ] Ensure compatibility with the existing `compile_scorch()` / `fit_scorch()` workflow
* [ ] Document the need for full-batch training for the standard Cox objective
* [ ] Add tests against `survival::coxph()`
* [ ] Add a small DeepSurv example or vignette

This would allow essentially any scorcher architecture with a single linear output to be trained as a DeepSurv model, while preserving scorcher's existing composable architecture rather than introducing a survival-specific model class.
